src/Simulations.jl

```diff
@@ -0,0 +1,174 @@
+module Simulations
+
+export Simulation, run!
+
+import Base: show
+
+using OrderedCollections: OrderedDict
+using Oceananigans: AbstractDiagnostic, AbstractOutputWriter
+
+using Oceananigans.Models
+using Oceananigans.Diagnostics
+using Oceananigans.OutputWriters
+using Oceananigans.TimeSteppers
+using Oceananigans.Utils
+
+mutable struct Simulation{M, Δ, C, I, T, W, R, D, O, P, F}
+                 model :: M
+                    Δt :: Δ
+         stop_criteria :: C
+        stop_iteration :: I
+             stop_time :: T
+       wall_time_limit :: W
+              run_time :: R
+           diagnostics :: D
+        output_writers :: O
+              progress :: P
+    progress_frequency :: F
+end
+
+"""
+    Simulation(model; Δt,
+         stop_criteria = Function[iteration_limit_exceeded, stop_time_exceeded, wall_time_limit_exceeded],
+        stop_iteration = Inf,
+             stop_time = Inf,
+       wall_time_limit = Inf,
+           diagnostics = OrderedDict{Symbol, AbstractDiagnostic}(),
+        output_writers = OrderedDict{Symbol, AbstractOutputWriter}(),
+              progress = nothing,
+    progress_frequency = 1)
+
+Construct an Oceananigans.jl `Simulation` for a `model` with time step `Δt`.
+
+Keyword arguments
+=================
+- `Δt`: Required keyword argument specifying the simulation time step. Can be a `Number`
+  for constant time steps or a `TimeStepWizard` for adaptive time-stepping.
+- `stop_criteria`: A list of functions (each taking a single argument, the `simulation`).
+  If any of the functions return `true` when the stop criteria is evaluated the simulation
+  will stop.
+- `stop_iteration`: Stop the simulation after this many iterations.
+- `stop_time`: Stop the simulation once this much model clock time has passed.
+- `wall_time_limit`: Stop the simulation if it's been running for longer than this many
+   seconds of wall clock time.
+- `progress`: A function with a single argument, the `simulation`. Will be called every
+  `progress_frequency` iterations. Useful for logging simulation health.
+- `progress_frequency`: How often to update the time step, check stop criteria, and call
+  `progress` function (in number of iterations).
+"""
+function Simulation(model; Δt,
+        stop_criteria = Function[iteration_limit_exceeded, stop_time_exceeded, wall_time_limit_exceeded],
```

Comment:
Since `Function` is an abstract type anyways, it might be helpful simply to allow this to be `Array{Any}`, so that users can add custom callable objects to it without having to subtype `Function`. But really not necessary for this PR at all.

Reply:
Small change but might make a user happy. Done.